05_debug/debug_module.c

```diff
@@ -0,0 +1,299 @@
+
+	sysfs_kobject = kobject_create_and_add(SYSFS_DIR_NAME, kernel_kobj);
+	if (sysfs_kobject == NULL) {
+		pr_err("%s: error creating kobject\n", __func__);
+		return -ENOMEM;
+	}
+
+	ret = sysfs_create_file(sysfs_kobject, &kobj_attr_sysfs.attr);
+	if (ret)
+		goto failed_sysfs_init;
+
+	ret = sysfs_create_file(sysfs_kobject, &kobj_attr_stat.attr);
+	if (ret)
+		goto failed_sysfs_init;
+
+
+	memset(&sysfs_conv, 0, sizeof(sysfs_conv));
+
+	pr_info("%s: success\n", __func__);
+	return 0;
+
+failed_sysfs_init:
+	pr_err("%s: error creating sysfs class attribute\n", __func__);
+	kobject_del(sysfs_kobject);
+	return ret;
+}
+
+static ssize_t stat_read(struct file *file,
+		char __user *pbuf, size_t count, loff_t *ppos)
+{
+	int res;
+	char stat_buff[PROCFS_BUFF_SIZE];
+	size_t to_copy = sprintf(stat_buff, "uppercase convertion stat:\n");
+
+	res = print_stat(&stat_buff[to_copy], &procfs_conv);
+	if (res < 0)
+		return res;
+
+	return simple_read_from_buffer(pbuf, count, ppos,
+					stat_buff, strlen(stat_buff));
+}
+
+static ssize_t procfs_read(struct file *file, char __user *pbuf,
+		size_t count, loff_t *ppos)
+{
+	if (procfs_conv.str_buff == NULL) {
+		pr_err("%s: no buffer to output\n", __func__);
+		return 0;
+	}
+
+	return simple_read_from_buffer(pbuf, count, ppos,
+			procfs_conv.str_buff, strlen(procfs_conv.str_buff));
+
+}
+
+static ssize_t procfs_write(struct file *file, const char __user *pbuf,
+		size_t count, loff_t *ppos)
+{
+	int res;
+	ssize_t not_copied, copied;
+	static char input_buf[PROCFS_BUFF_SIZE];
+
+	kzfree(procfs_conv.str_buff);
+
+	not_copied = copy_from_user(input_buf, pbuf, count);
+	copied = count - not_copied;
+
+	procfs_conv.str_buff = kstrdup(input_buf, GFP_KERNEL);
+
+	res = convert_case(false, &procfs_conv, copied);
+	if (res < 0)
+		return res;
+
+	return copied;
+}
+
+static struct proc_dir_entry *proc_dir;
+static struct proc_dir_entry *ent_uppercase;
+static const struct file_operations procfs_ops = {
+	.owner = THIS_MODULE,
+	.read = procfs_read,
+	.write = procfs_write,
+};
+static const struct file_operations stat_ops = {
+	.owner = THIS_MODULE,
+	.read = stat_read
+};
+
+int init_procfs(void)
+{
+	proc_dir = proc_mkdir(PROCFS_DIR_NAME, NULL);
+	if (proc_dir == NULL) {
+		pr_err("%s: error creating procfs directory\n", __func__);
+		return -ENOMEM;
+	}
+
+	ent_uppercase = proc_create(PROCFS_ENT_NAME, 0666,
+					proc_dir, &procfs_ops);
+	if (ent_uppercase == NULL)
+		goto failed_procfs_init;
+
+	ent_uppercase = proc_create(STAT_ENT_NAME, 0444, proc_dir, &stat_ops);
+	if (ent_uppercase == NULL)
+		goto failed_procfs_init;
+
+	memset(&procfs_conv, 0, sizeof(procfs_conv));
+
+	pr_info("%s: success\n", __func__);
+	return 0;
+
+failed_procfs_init:
+	pr_err("%s: error creating procfs entry\n", __func__);
+	proc_remove(proc_dir);
+	return -ENOMEM;
+}
+
+static int debug_module_init(void)
+{
+	int ret;
+
+	pr_info("%s: module starting\n",  __func__);
+
+	kzfree(procfs_conv.str_buff);
+	kzfree(sysfs_conv.str_buff);
```

Comment:
Wrong place. Did you want to place them into exit function?

Reply:
Yes. My mistake, will fix